getappmap/appmap-node
# feat: Async context tracking

### Implement Async Tracking in AppMap-Node Agent Using AsyncLocalStorage

#### Problem

The current implementation of the AppMap-Node agent struggles with proper asynchronous tracking, leading to misleading call hierarchies. Specifically, the agent does not maintain the correct context of asynchronous computations, resulting in inaccuracies when representing operations that involve `Promise` or other asynchronous constructs.

For example:


```javascript
function bar(x) { return x * 2; }

async function foo() {
  await setTimeout(() => 1, 100);
  console.log(bar(1));
}

foo();
console.log(bar(2));

```
In this case, the call to `bar(1)` is incorrectly represented as coming after `bar(2)` because the `foo()`'s context is lost early.

#### Analysis

The main issues with the current implementation are:

1.  **Fixup After Promise Settlements**: Return events are only fixed up when `Promise` settlements are observed.
2.  **Early Context Closure**: The context of an async function closes as soon as it returns, leading to incorrect parent-child relationships in the reported events.
3.  **Immediate Event Streaming**: Events are streamed as they come, which can disrupt the correct chronological context, especially with async functions.

We need to update the tracking mechanism to keep the context alive throughout the async function's execution and ensure that events are accurately reported.

#### Proposed Changes

1.  **Context Preservation**:
    
    *   Use `AsyncLocalStorage` to preserve the execution context across asynchronous operations.
    *   Ensure any asynchronous operations, such as `await` or callback procedures, maintain the correct parent context in the generated AppMaps.
2.  **Event Streaming**:
    
    *   Amend the event streaming logic to respect asynchronous boundaries instead of streaming events immediately.
    *   Queue events until the corresponding async context finalizes, ensuring the correct call hierarchy is maintained.

#### Detailed Changes

1.  **Update `Recording.ts`**:
    
    *   Modify the function that handles function calls to use `AsyncLocalStorage` to store context information.
    *   Adjust the logic that generates function return and exception events to account for async contexts.
2.  **Adjust Event Generation Logic**:
    
    *   Ensure that the `fixupPromise` function correctly associates the context of async operations.
    *   Keep async contexts open until the async function fully completes.
3.  **Modify Event Emission**:
    
    *   Adjust how and when events are emitted to ensure they respect the actual call hierarchy within asynchronous flows.
4.  **Tests and Validation**:
    
    *   Update the relevant tests in `test/jest/asyncLib.test.js` and other relevant test files to validate the accuracy of the reported call hierarchies.
    *   Ensure extensive tests cover various async scenarios to validate the new tracking mechanism's effectiveness.

Making these adjustments will significantly improve the accuracy of asynchronous tracking in the AppMap-Node agent, providing clearer and more accurate call hierarchies and execution flows in AppMaps.
